flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -503,6 +503,11 @@ private void runMailboxLoop() throws Exception {
 				throw e;
 			}
 		}
+		catch (CancelTaskException e) {
+			// inputs closed before the mail loop has been interrupted or gracefully terminated
+			// swallow exception here but set canceled flag, which will trigger a new CancelTaskException in #invoke()
+			canceled = true;
```

Comment:
```
if (!canceled) { LOG.error("Received CancelTaskException while we are not canceled. This is a bug and should be reported", e); }
```
?

Reply:
> Can we add some test coverage for that?

Let me try that.

Reply:
> I would prefer the cancellation mechanism to stay clean and `canceled` being set to `true` only in one place, without using exceptions for that.

Understood and I'd ideally would do that, but we already have the situation that the gates are closed before `#cancel()` has been called (or else the exception would have been swallowed by your commit).
Then, the whole cleanup of Task#invoke may happen without canceled being set at all and I wouldn't be surprised to see the next issue arising.

To further emphasize, if you check `StreamTask#invoke` you'd see that if `canceled` is set, we'd skip `afterInvoke` and also take the exceptional code path. If we just swallow the exception and do not set canceled, we'd go a whole other code path.